### ISSUE TITLE
Change mipsel-musl to soft float

### ISSUE THREAD

**Issue by alexcrichton** [rust-lang/rust#34922]
It was [pointed out](https://github.com/rust-lang/rust/pull/34910#issuecomment-233694689) by @japaric that the mipsel-musl toolchain should use soft float like the mips-musl toolchain as almost all devices it targets all have soft float. Changing this will require a few pieces:

* [x] Change the [linux-cross docker image](https://github.com/rust-lang/rust-buildbot/blob/master/slaves/linux-cross/Dockerfile) in how it builds the [mipsel-linux-musl toolchain](https://github.com/rust-lang/rust-buildbot/blob/0a0b6d681789d8fabdfac023177ef863ae946f1b/slaves/linux-cross/Dockerfile#L84) by [following the instructions](https://github.com/rust-lang/rust-buildbot/blob/master/slaves/linux-cross/README.md) to regenerate the [config file](https://github.com/rust-lang/rust-buildbot/blob/master/slaves/linux-cross/mipsel-linux-musl.config). (anyone can do this, feel free to send a PR) - https://github.com/rust-lang/rust-buildbot/pull/108
* [x] Publish a new linux-cross docker image (I can do this)
* [x] Restart buildbot to pick up the new linux-cross docker image (I can do this)
* [ ] Send a PR to add `+soft-float` to the target-features of [the mipsel-musl target spec](https://github.com/rust-lang/rust/blob/27e766d7bc84be992c8ddef710affc92ef4a0adf/src/librustc_back/target/mipsel_unknown_linux_musl.rs#L25), like the [mips-musl target spec](https://github.com/rust-lang/rust/blob/27e766d7bc84be992c8ddef710affc92ef4a0adf/src/librustc_back/target/mips_unknown_linux_musl.rs#L25)

I'm gonna tag this as E-easy as it's pretty clear how to proceed, and I can help out anyone willing to do so! You even get to play around with our infrastructure a bit :)

**Comment by nagisa** [comment]
Just to mirror what I [said] in the PR which spawned this issue, I’m against changing this target to soft-float before we collect information on what ABI conventional linux distributions, and not OpenWRT, use.

[said]: https://github.com/rust-lang/rust/pull/34910#issuecomment-233739886

**Comment by nagisa** [comment]
For what its worth, the Debian MIPS ports are all hard-float.

**Comment by alexcrichton** [comment]
@nagisa perhaps yeah, but I believe that OpenWRT/@japaric are the primary users of mips-musl right now, as in we don't actually have any demand for mips-unknown-linux-musl that's *not* openwrt that I know of.

**Comment by alexcrichton** [comment]
Ok, I've published `alexcrichton/rust-slave-linux-cross:2016-07-19`, the next step is to update it on the bots. I unfortunately *just* restarted the bots this morning due to some corrupt images. I'd like to try to land https://github.com/rust-lang/rust/pull/34873 first to help along the LLVM upgrade to unblock MIR, but after that I can try to restart.

**Comment by japaric** [comment]
> For what its worth, the Debian MIPS ports are all hard-float.

@nagisa, are those ports musl based or glibc based?

**Comment by nagisa** [comment]
@japaric while Debian itself uses `glibc`, as the example of `x86_64-unknown-linux-musl` demonstrates, the `-musl` targets are better known as the targets which produce fully static binaries, rather than targets which target a certain kind of distribution.

I guess since musl is built-in and linked to statically, either soft and hard float will be correct… as long as the binary doesn’t link to any other library from the system. At that point the actual libc the distro is based on ends up not mattering much, but the FP ABI does.

**Comment by japaric** [comment]
@nagisa Ohh, I thought you were suggesting running `mips-musl` Rust binaries on a MIPS device running Debian.

> the -musl targets are better known as the targets which produce fully static binaries

The `mips(el)-musl` Rust binaries are **not** statically linked though; they are dynamically linked to libc like pretty much every other target. Statically linked Rust binaries for MIPS are not currently possible because those binaries must be statically linked to libunwind, and `libunwind` doesn't support MIPS right now (or least it didn't last month -- I haven't been following its development closely).

**Comment by nagisa** [comment]
> Ohh, I thought you were suggesting running mips-musl Rust binaries on a MIPS device running Debian.

Yes, I was. I had produced a `-musl` binaries for various architectures that were supposed to run on some sort of Debian/Fedora/Arch in the past, and I’d expect a binary targeting `mips-unknown-linux-musl` to work on a generic MIPS distro such as Debian too. Making `musl` targets soft-float essentially breaks that use case.

I’d rather make my own `mips-unknown-linux-musl` target (just like I need to with ARM currently) when I need one and have rustc provide a `mips-openwrt-linux-musl` instead, in order to avoid the confusion.

TL;DR: pretty much all I’m unhappy about is the naming of soft-fp MIPS targets.

> libunwind doesn't support MIPS right now (or least it didn't last month -- I haven't been following its development closely).

Which makes

> … as long as the binary doesn’t link to any other library from the system. At that point the actual libc the distro is based on ends up not mattering much, but the FP ABI does.

FP ABI used matter all, rather than some of the time.

**Comment by alexcrichton** [comment]
Ok, buildbot is restarted, next patch should be good to land!
